007_buildkite_dynamic_pipelines/README.md

```diff
@@ -0,0 +1,298 @@
+
+In the Buildkite web UI, create a new build for the pipeline. Check that it
+still completes successfully, and goes green.
+
+If the updated pipeline doesn't go green you'll need to diagnose and fix this
+before continuing.
+
+### Add some files to the repository
+
+#### :arrow_right: Create a location for your CUE files
+
+At the top-level of your repository, create a directory to hold your Buildkite
+CUE files:
+
+:computer: `terminal`
+```sh
+cd $(git rev-parse --show-toplevel)   # make sure we're sitting at the repository root
+mkdir -p internal/ci/buildkite/
+```
+
+| :grey_exclamation: Info :grey_exclamation: |
+|:------------------------------------------ |
+| You *can* change the `buildkite` directory's *location*, but **don't change its name**.<br>If you change its location, you'll have to adapt some commands to match the new location, as you follow the rest of this guide.
+
+#### :arrow_right: Create CUE tool
+
+:floppy_disk: `internal/ci/buildkite/dynamic_tool.cue`
+```CUE
+package pipeline
+
+import (
+	"encoding/yaml"
+	"tool/os"
+	"tool/cli"
+)
+
+ENV: command.emit_dynamic_steps.env
+dynamic_steps: [...]
+
+command: emit_dynamic_steps: {
+	env:  os.Environ
+	emit: cli.Print & {
+		text: yaml.Marshal({steps: dynamic_steps})
+	}
+}
+```
+
+#### :arrow_right: Create dynamic steps
+
+:floppy_disk: `internal/ci/buildkite/dynamic.steps.cue`
+```CUE
+package pipeline
+
+ENV: {
+	BUILDKITE_AGENT_ENDPOINT: *"test value" | _
+	BUILDKITE_BRANCH:         *"test value" | _
+}
+dynamic_steps: [
+	{command: "echo dynamic step 1"},
+	{command: "echo dynamic step 2"},
+	{command: "echo The agent endpoint in use is \(ENV.BUILDKITE_AGENT_ENDPOINT)"},
+	{command: "echo dynamic step 3"},
+	if ENV.BUILDKITE_BRANCH == "main" {"wait"},
+	if ENV.BUILDKITE_BRANCH == "main" {
+		{
+			command: "echo Deploy!"
+			label:   ":rocket:"
+		}
+	},
+]
+```
+
+These steps are (of course!) just examples. Replace them with steps that meet
+your pipeline's requirements. You can use any CUE features and structure,
+including conditionally using the contents and values of the context-specific
+`ENV` struct, so long as the `dynamic_steps` field ultimately contains an
+ordered list of your steps.
+
+#### :arrow_right: Create CUE schema
+
+:floppy_disk: `internal/ci/buildkite/dynamic.schema.cue`
+```CUE
+package pipeline
+
+ENV: close({[string]: string})
+#Step: _
+dynamic_steps: [...#Step]
+```
+
+|  :grey_exclamation: Info :grey_exclamation: |
+|:------------------------------------------- |
+| It would be great if we could use [Buildkite's authoritative pipeline schema](https://github.com/buildkite/pipeline-schema) here. Unfortunately, CUE's JSONSchema support can't currently import it. This is being tracked in CUE Issues [#2698](https://github.com/cue-lang/cue/issues/2698) and [#2699](https://github.com/cue-lang/cue/issues/2699), and this guide should be updated once the schema is useable.
+
+#### :arrow_right: Create CUE policy
```

Comment:
I think we need some motivate on why/when people would write policy here. I'm also not clear you are writing a policy here, because you appear to be augmenting the schema. In reality, that schema would be imported so you would augment it in this way. Hence I would write it as pattern constraints.

Reply:
Done.